pkg/codegen/internal/tstypes/tstypes_test.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2016-2021, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tstypes
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/davecgh/go-spew/spew"
+
+	"github.com/stretchr/testify/assert"
+	"pgregory.net/rapid"
+)
+
+func TestParenInsert(t *testing.T) {
```

Comment:
Admittedly verbose but foolproof. Works on a simplified grammar but can be easily extended if we run into more grammar peculiarities. This checks that unparsing and auto-parenthesizing the AST into tokens gets parsed back into the same ast.